File: src/js/4_BehavioralPatterns/4_Iterator/tree_traversal.js
class Node
{
  constructor(value, left=null, right=null)
  {
    this.value = value;
    this.left = left;
    this.right = right;
    this.parent = null;

    if (this.left)
      left.parent = this;
    if (this.right)
      right.parent = this;
  }
}

function makeInOrderIterator(root)
{
  // go to leftmost
  let current = root;
  while (current.left) {
    current = current.left;
  }
  let yieldedStart = false;

  return {
    next: function() {
      if (!yieldedStart)
      {
        yieldedStart = true;
        return {
          value: current,
          done: false
        };
      }
      if (current.right)
      {
        current = current.right;
        while (current.left)
        {
          current = current.left;
        }
        return {
          value: current,
          done: false
        };
      }
      else
      {
        let p = current.parent;
        while (p && current === p.right)
        {
          current = p;
          p = p.parent;
        }
        current = p;
        return {
          value: current,
          done: current == null
        };
      }
    }, // next

    // this makes the iterator iterable
    [Symbol.iterator]: function() { return this; }
  };
}

class BinaryTree
{
  constructor(rootNode)
  {
    this.rootNode = rootNode;
  }

  // assuming only one form of iteration
  [Symbol.iterator]()
  {
    return makeInOrderIterator(this.rootNode);
  }

  * betterInOrder()
  {
    function* traverse(current)
    {
      if (current.left)
      {
        for (let left of traverse(current.left))
          yield left;
      }
      yield current;
      if (current.right)
      {
        for (let right of traverse(current.right))
          yield right;
      }
    }
    for (let node of traverse(this.rootNode))
      yield node;
  }

  get inOrder()
  {
    return makeInOrderIterator(this.rootNode);
  }
}

//   1
//  / \
// 2   3

// in-order:  213
// preorder:  123
// postorder: 231

let root = new Node(1, new Node(2), new Node(3));

// c++ style
let it = makeInOrderIterator(root);
let result = it.next();
while (!result.done)
{
  console.log(result.value.value);
  result = it.next();
}

let tree = new BinaryTree(root);

for (let x of tree)
  console.log(x.value);

console.log([...tree].map(x => x.value));

console.log([...tree.inOrder].map(x => x.value));

// a generator is both an iterator and iterable
console.log('using a generator...');
console.log([...tree.betterInOrder()].map(x => x.value));

for (let x of tree.betterInOrder())
  console.log(x.value);

File: src/swift/4_BehavioralPatterns/4_Iterator/TreeTraversal.swift
import Foundation
import XCTest

class Node<T>
{
  var value: T
  var left: Node<T>? = nil
  var right: Node<T>? = nil
  var parent: Node<T>? = nil

  init(value: T)
  {
    self.value = value
  }

  init(value: T, left: Node<T>, right: Node<T>)
  {
    self.value = value
    self.left = left
    self.right = right

    left.parent = self
    right.parent = self
  }
}

class InOrderIterator<T> : IteratorProtocol
{
  var current: Node<T>?
  var root: Node<T>
  var yieldedStart = false

  init(root: Node<T>)
  {
    self.root = root
    current = root
    while (current!.left != nil)
    {
      current = current!.left!
    }
  }

  func reset()
  {
    current = root
    yieldedStart = false // really?
  }

  func next() -> Node<T>?
  {
    if !yieldedStart
    {
      yieldedStart = true
      return current
    }

    if current!.right != nil
    {
      current = current!.right
      while current!.left != nil
      {
        current = current!.left
      }
      return current
    }
    else
    {
      var p = current!.parent
      while (p != nil) && (current === p!.right)
      {
        current = p!
        p = p!.parent
      }
      current = p
      return current
    }
  }
}

class BinaryTree<T> : Sequence
{
  private let root: Node<T>

  init(root: Node<T>)
  {
    self.root = root
  }

  func makeIterator() -> InOrderIterator<T>
  {
    return InOrderIterator<T>(root: root)
  }
}

func main()
{
  //   1
  //  / \
  // 2   3

  // in-order:  213
  // preorder:  123
  // postorder: 231

  let root = Node(
    value: 1,
    left: Node(value: 2),
    right: Node(value: 3))

  // direct use of the iterator (C++ style)
  let it = InOrderIterator(root: root)
  while let element = it.next()
  {
    print(element.value, terminator: " ")
  }
  print()

  // feel free to stick it into a sequence
  let nodes = AnySequence{InOrderIterator(root: root)}
  print(nodes.map{ $0.value })

  // or, if you want to make an entire class...
  let tree = BinaryTree(root: root)

  print(tree.map{ $0.value })
}

main()